contextualized/dags/lightning_modules.py

```diff
@@ -289,23 +337,33 @@ def training_epoch_end(self, training_step_outputs, logs=None):
         )
         epoch_dag_loss = 0
         for ret in training_step_outputs:
-            batch_dag_loss = dag_loss(
+            batch_dag_loss = self.ss_dag_loss(
                 self.predict_step(ret["train_batch"], ret["train_batch_idx"]),
-                self.ss_alpha,
-                self.ss_rho,
+                **self.ss_dag_params,
             ).detach()
             epoch_dag_loss += (
                 len(ret["train_batch"][0]) / epoch_samples * batch_dag_loss
             )
+        self.ss_dag_params = self._maybe_update_alpha_rho(
+            epoch_dag_loss, self.ss_dag_params
+        )
+        self.arch_dag_params = self._maybe_update_alpha_rho(
+            epoch_dag_loss, self.arch_dag_params
+        )
+
+    def _maybe_update_alpha_rho(self, epoch_dag_loss, dag_params):
         if (
-            self.use_dynamic_alpha_rho
-            and epoch_dag_loss > self.tol * self.h_old
-            and self.ss_alpha < 1e12
-            and self.ss_rho < 1e12
+            dag_params.get("use_dynamic_alpha_rho", False)
+            and epoch_dag_loss > dag_params["tol"] * dag_params["h_old"]
+            and dag_params["alpha"] < 1e12
+            and dag_params["rho"] < 1e12
         ):
-            self.ss_alpha = self.ss_alpha + self.ss_rho * epoch_dag_loss
-            self.ss_rho = self.ss_rho * 1.1
-        self.h_old = epoch_dag_loss
+            dag_params["alpha"] = (
+                dag_params["alpha"] + dag_params["rho"] * epoch_dag_loss
+            )
+            dag_params["rho"] = dag_params["rho"] * 1.1
```

Comment:
I think we might want to add the update "rate" (1.1 by default) as a loss parameter. This seems to have interplay with early stopping parameter choices.

Reply:
Resolved